### PR TITLE
Fix debuginfo for pointers/references to unsized types

### LINKED ISSUE #92718: Incomplete/wrong debuginfo for references and pointers to slice-like unsized types
References and pointers to unsized types are "fat" which means that they hold not just the address of the pointee but also an additional field that contains either the size of the pointee's unsized field (for slice-like pointees) or the addr of the vtable (for trait object pointers). 

The compiler generates debuginfo for this additional field in most cases (e.g. for `&[T]` and `*mut [T]`) -- but if the pointee is an unsized struct then the compiler generates regular thin pointer debuginfo which does not describe the additional field.

Example:

```rust
pub struct SomeUnsizedType {
    pub a: u32,
    pub b: u64,
    pub c: [u8],
}

pub fn references(_slice_ref: &[u64], _struct_ref: &SomeUnsizedType) {}
```

The debuginfo for `&[u64]` looks as expected:

```txt
// the fat reference &[u64] is described as a struct with two fields,
// `data_ptr` and `length`.
< 1><0x0000012c>    DW_TAG_structure_type
                      DW_AT_name                  &[u64]
                      DW_AT_byte_size             0x00000010
                      DW_AT_alignment             0x00000008
< 2><0x00000133>      DW_TAG_member
                        DW_AT_name                  data_ptr
                        DW_AT_type                  <0x000000e3>
                        DW_AT_alignment             0x00000008
                        DW_AT_data_member_location  0
< 2><0x0000013e>      DW_TAG_member
                        DW_AT_name                  length
                        DW_AT_type                  <0x000000f7>
                        DW_AT_alignment             0x00000008
                        DW_AT_data_member_location  8
```

But `&SomeUnsizedType` does not have the additional field:

```txt
< 1><0x0000014a>    DW_TAG_pointer_type
                      DW_AT_type                  <0x0000009b>
                      DW_AT_name                  &SomeUnsizedType
                      DW_AT_address_class         0x00000000

< 2><0x0000009b>      DW_TAG_structure_type
                        DW_AT_name                  SomeUnsizedType
                        DW_AT_byte_size             0x00000010
                        DW_AT_alignment             0x00000008
< 3><0x000000a2>        DW_TAG_member
                          DW_AT_name                  a
                          DW_AT_type                  <0x0000010b>
                          DW_AT_alignment             0x00000004
                          DW_AT_data_member_location  8
< 3><0x000000ad>        DW_TAG_member
                          DW_AT_name                  b
                          DW_AT_type                  <0x000000f0>
                          DW_AT_alignment             0x00000008
                          DW_AT_data_member_location  0
< 3><0x000000b8>        DW_TAG_member
                          DW_AT_name                  c
                          DW_AT_type                  <0x00000112>
                          DW_AT_alignment             0x00000001
                          DW_AT_data_member_location  12

// This is the type of the unsized field `c`
< 1><0x00000112>    DW_TAG_array_type
                      DW_AT_type                  <0x0000011e>
< 2><0x00000117>      DW_TAG_subrange_type
                        DW_AT_type                  <0x00000125>
                        DW_AT_lower_bound           0x00000000
< 1><0x0000011e>    DW_TAG_base_type
                      DW_AT_name                  u8
                      DW_AT_encoding              DW_ATE_unsigned
                      DW_AT_byte_size             0x00000001
< 1><0x00000125>    DW_TAG_base_type
                      DW_AT_name                  __ARRAY_SIZE_TYPE__
                      DW_AT_byte_size             0x00000008
                      DW_AT_encoding              DW_ATE_unsigned

```

The last field is described as a zero-sized array `[u8; 0]`. As a consequence, a debugger has no way of finding out how many elements the field actually holds. This affects a number of common types from the standard library, like `Rc`, `Arc`, `CStr` which all use unsized structs internally.
 
### Meta
`rustc --version --verbose`:
```
rustc 1.59.0-nightly (8f3238f89 2022-01-02)
binary: rustc
commit-hash: 8f3238f898163f09726c3d2b2cc9bafb09da26f3
commit-date: 2022-01-02
host: x86_64-unknown-linux-gnu
release: 1.59.0-nightly
LLVM version: 13.0.0
```

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR makes the compiler emit fat pointer debuginfo in all cases. Before, we sometimes got thin-pointer debuginfo, making it impossible to fully interpret the pointed to memory in debuggers. The code is actually cleaner now, especially around generation of trait object pointer debuginfo.

Fixes https://github.com/rust-lang/rust/issues/92718

~~Blocked on https://github.com/rust-lang/rust/pull/92729.~~

### CODE REVIEW

**Comment by michaelwoerister** [comment]
This PR makes type names, that are part of a pointer type, fully qualified. For example, `&Foo<u8>` becomes `&my_crate::some_mod::Foo<u8>`. This is the correct behavior, but it occurs to me that that might have performance implications. 

@bors try @rust-timer queue

**Comment by oli-obk** [comment]
This lgtm from a pure impl perspective, but I think someone who knows llvm should look over it

r? rust-lang/wg-llvm

**Comment by michaelwoerister** [comment]
There doesn't seem to be an LLVM group in https://github.com/rust-lang/highfive/blob/master/highfive/configs/rust-lang/rust.json yet.

r? @davidtwco, who knows debuginfo well (but feel free to re-assign if you don't want to review!)

File: compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs [hunk 1]
```diff
@@ -495,83 +555,57 @@ fn subroutine_type_metadata<'ll, 'tcx>(
...
 }
 
-// FIXME(1563): This is all a bit of a hack because 'trait pointer' is an ill-
-// defined concept. For the case of an actual trait pointer (i.e., `Box<Trait>`,
-// `&Trait`), `trait_object_type` should be the whole thing (e.g, `Box<Trait>`) and
-// `trait_type` should be the actual trait (e.g., `Trait`). Where the trait is part
-// of a DST struct, there is no `trait_object_type` and the results of this
-// function will be a little bit weird.
-fn trait_pointer_metadata<'ll, 'tcx>(
+// Create debuginfo for `dyn SomeTrait` types. Currently these are empty structs
```
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs:558]:
```suggestion
/// Create debuginfo for `dyn SomeTrait` types. Currently these are empty structs
```
nit: documentation comment (need to change the line below this too)

File: compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs [hunk 2]
```diff
@@ -495,83 +555,57 @@ fn subroutine_type_metadata<'ll, 'tcx>(
...
+        composite_type_metadata(cx, dyn_type, &type_name, unique_type_id, vec![], NO_SCOPE_METADATA)
+    } else {
+        bug!("Only ty::Dynamic is valid for dyn_type_metadata(). Found {:?} instead.", dyn_type)
+    }
+}
 
-        // No object type, use the trait type directly (no scope here since the type
-        // will be wrapped in the dyn$ synthetic type).
-        None => (NO_SCOPE_METADATA, compute_debuginfo_type_name(cx.tcx, trait_type, true)),
+// Create debuginfo for `[T]` and `str`. These are unsized.
```
Comment (davidtwco) [inline compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs:573]:
```suggestion
/// Create debuginfo for `[T]` and `str`. These are unsized.
```
nit: documentation comment (need to change rest of the comment too)

**Review by davidtwco — CHANGES_REQUESTED** [review]
This looks good to me, two small nits but otherwise I think this is good.

**Comment by michaelwoerister** [comment]
Thanks for the reviews! Comments should be addressed now.

**Comment by jackh726** [comment]
There weren't any perf regressions and I don't think this needs anything special in regards to rollups.

@bors rollup=maybe

File: compiler/rustc_codegen_llvm/src/lib.rs [hunk 3]
```diff
@@ -9,6 +9,7 @@
 #![feature(crate_visibility_modifier)]
 #![feature(extern_types)]
 #![feature(nll)]
+#![feature(let_else)]
```
Comment (matthiaskrgr) [inline compiler/rustc_codegen_llvm/src/lib.rs:12]: It seems to have already been added by another PR in the meantime causing duplicate feature warnings :(
https://github.com/rust-lang/rust/pull/93218#issuecomment-1019379001

**Comment by michaelwoerister** [comment]
> There weren't any perf regressions and I don't think this needs anything special in regards to rollups.

The `vec-slices` debuginfo test has been ignored for a long time and there might still be something GDB-version dependent in there, so let's exclude this PR from rollups.

@bors rollup=never

**Comment by michaelwoerister** [comment]
Fixed the duplicate feature warning.

@bors r=davidtwco,oli-obk

**Comment by michaelwoerister** [comment]
Let's see if the tests pass for old GDB versions too now.

@bors r=davidtwco,oli-obk

**Comment by michaelwoerister** [comment]
Fixed the test for 32-bit MSVC.

@bors r=davidtwco,oli-obk

**Comment by michaelwoerister** [comment]
Added missing `[...]` at the end of some CDB check lines.

@bors r=davidtwco,oli-obk

**Comment by michaelwoerister** [comment]
Fix doc-comment so it's not interpreted as doctest.

@bors r=davidtwco,oli-obk

**Comment by michaelwoerister** [comment]
@rust-lang/infra This is the second time in this PR that there's a core dump in compiletest after mir-opt testing: https://github.com/rust-lang/rust/runs/4978717078?check_suite_focus=true#step:26:3567
I was not able to reproduce the crash locally. Is this a known issue?

**Comment by michaelwoerister** [comment]
I was able to reproduce the LLDB test failures on Linux. The last commit removes the `DW_AT_artifical` attribute from fat pointer fields (as was the case before this PR). That allows LLDB to see the fields again.

@bors r=davidtwco,oli-obk

**Pull request merged**